instrumentation/github.com/Shopify/sarama/consumer.go

```diff
@@ -0,0 +1,122 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sarama
+
+import (
+	"context"
+
+	"github.com/Shopify/sarama"
+
+	"go.opentelemetry.io/otel/api/kv"
+	"go.opentelemetry.io/otel/api/propagation"
+	"go.opentelemetry.io/otel/api/standard"
+	"go.opentelemetry.io/otel/api/trace"
+)
+
+type partitionConsumer struct {
+	sarama.PartitionConsumer
+	messages chan *sarama.ConsumerMessage
+}
+
+// Messages returns the read channel for the messages that are returned by
+// the broker.
+func (pc *partitionConsumer) Messages() <-chan *sarama.ConsumerMessage {
+	return pc.messages
+}
+
+// WrapPartitionConsumer wraps a sarama.PartitionConsumer causing each received
+// message to be traced.
+func WrapPartitionConsumer(serviceName string, pc sarama.PartitionConsumer, opts ...Option) sarama.PartitionConsumer {
+	cfg := newConfig(serviceName, opts...)
+
+	wrapped := &partitionConsumer{
+		PartitionConsumer: pc,
+		messages:          make(chan *sarama.ConsumerMessage),
+	}
+	go func() {
+		msgs := pc.Messages()
+
+		var prevSpan trace.Span
+		for msg := range msgs {
+			// Extract a span context from message to link.
+			carrier := NewConsumerMessageCarrier(msg)
+			parentSpanContext := trace.RemoteSpanContextFromContext(propagation.ExtractHTTP(context.Background(), cfg.Propagators, carrier))
```

Comment:
I'm not sure this is necessary.  If you use the `context.Context` returned from `propagation.ExtractHTTP()` in your call to `cfg.Tracer.Start()` it will determine that the context contains a remote span context and correctly associate it as a child with a remote parent.

The current implementation will, I believe, result in a new trace graph with the root linked to the parent span context, which may not be the desired behavior.